libcodechecker/analyze/plist_parser.py

```diff
@@ -184,3 +184,117 @@ def parse_plist(path):
         LOG.error(ex)
     finally:
         return files, reports
+
+
+def fids_in_range(rng):
+    """
+    Get the file ids from a range.
+    """
+    fids = []
+    for r in rng:
+        for l in r:
+            fids.append(l['file'])
+    return fids
+
+
+def fids_in_edge(edges):
+    """
+    Get the file ids from an edge.
+    """
+    fids = []
+    for e in edges:
+        start = e['start']
+        end = e['end']
+        for l in start:
+            fids.append(l['file'])
+        for l in end:
+            fids.append(l['file'])
+    return fids
+
+
+def fids_in_path(report_data, file_ids_to_remove):
+    """
+    Skip diagnostic sections and collect file ids in
+    report paths for the remaining diagnostic sections.
+    """
+    all_fids = []
+
+    kept_diagnostics = []
+
+    for i, diag in enumerate(report_data['diagnostics']):
+
+        if diag['location']['file'] in file_ids_to_remove:
+            continue
+
+        kept_diagnostics.append(diag)
+
+        for pe in diag['path']:
+            path_fids = []
+            try:
+                fids = fids_in_range(pe['ranges'])
+                path_fids.extend(fids)
+            except KeyError:
+                pass
+
+            try:
+                fid = pe['location']['file']
+                path_fids.append(fid)
+            except KeyError:
+                pass
+
+            try:
+                fids = fids_in_edge(pe['edges'])
+                path_fids.extend(fids)
+            except KeyError:
+                pass
+
+            all_fids.extend(path_fids)
+
+    return all_fids, kept_diagnostics
+
+
+def remove_report_from_plist(plist_content, skip_handler):
+    """
+    Parse the original plist content provided by the analyzer
+    and return a new plist content where reports were removed
+    if they should be skipped.
+    """
+    new_data = {}
+    report_data = plistlib.readPlistFromString(plist_content)
+
+    file_ids_to_remove = []
+
+    for i, f in enumerate(report_data['files']):
+        if skip_handler.should_skip(f):
+            file_ids_to_remove.append(i)
+
+    fids, kept_diagnostics = fids_in_path(report_data, file_ids_to_remove)
+    report_data['diagnostics'] = kept_diagnostics
+
+    remove_file = True
```

Comment:
What does this loop filter exactly? If we mark a file to be removed, but it is still used somewhere else?

Reply:
In the plist there is a `files` array with all the files where a report was found. We can safely remove a file from this list only if that file is not found in any of the other report paths.